File: client/modules/scanner-module/ios/ScannerModule.swift
import ExpoModulesCore

public class ScannerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ScannerModule")

    Events("onScanSuccess", "onError")

    AsyncFunction("requestCameraPermissionAsync") { () -> Bool in
      return true
    }

    Function("startScanning") {
    }

    Function("stopScanning") {
    }

    View(ScannerModuleView.self) {
      Prop("isScanning") { (view: ScannerModuleView, scanning: Bool) in
        view.setScanning(scanning)
      }
      Events("onScanSuccess", "onError")
    }
  }
}

File: apps/client/modules/scanner-module/android/src/main/java/expo/modules/scannermodule/ScannerModule.kt
package expo.modules.scannermodule

import android.Manifest
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class ScannerModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ScannerModule")

    Events("onScanSuccess", "onError")

    AsyncFunction("requestCameraPermissionAsync") { promise: Promise ->
      val context = appContext.reactContext
      if (context == null) {
        promise.resolve(false)
        return@AsyncFunction
      }
      val result = androidx.core.content.ContextCompat.checkSelfPermission(
        context,
        Manifest.permission.CAMERA
      )
      if (result == android.content.pm.PackageManager.PERMISSION_GRANTED) {
        promise.resolve(true)
        return@AsyncFunction
      }
      promise.resolve(false)
    }

    View(ScannerModuleView::class) {
      Prop("isScanning") { view: ScannerModuleView, scanning: Boolean ->
        view.setIsScanning(scanning)
      }
      Events("onScanSuccess", "onError")
    }
  }
}
